File: internal/sync/engine.go
package sync

import (
	"encoding/json"
	"log"
	"time"

	"github.com/linfree/open-todo/internal/store"
)

type Engine struct {
	store    *store.Store
	client   *Client
	lastSync time.Time
	stopCh   chan struct{}
}

func NewEngine(st *store.Store, serverURL, token string) *Engine {
	return &Engine{
		store:    st,
		client:   NewClient(serverURL, token),
		lastSync: time.Now().UTC(),
		stopCh:   make(chan struct{}),
	}
}

func (e *Engine) Start() {
	log.Println("[sync] engine started")
	go e.pushLoop()
	go e.pullLoop()
}

func (e *Engine) Stop() {
	close(e.stopCh)
}

func (e *Engine) pushLoop() {
	ticker := time.NewTicker(3 * time.Second)
	defer ticker.Stop()
	for {
		select {
		case <-e.stopCh:
			return
		case <-ticker.C:
			e.doPush()
		}
	}
}

func (e *Engine) pullLoop() {
	e.doPull()
	ticker := time.NewTicker(30 * time.Second)
	defer ticker.Stop()
	for {
		select {
		case <-e.stopCh:
			return
		case <-ticker.C:
			e.doPull()
		}
	}
}

func (e *Engine) doPush() {
	changes, err := e.store.GetUnsyncedChanges()
	if err != nil || len(changes) == 0 {
		return
	}

	var records []SyncRecord
	var ids []int64
	for _, c := range changes {
		data, err := e.store.GetFullRecord(c.TableName, c.RecordID)
		if err != nil {
			log.Printf("[sync] skip %s/%s: %v", c.TableName, c.RecordID, err)
			continue
		}
		records = append(records, SyncRecord{
			TableName: c.TableName,
			RecordID:  c.RecordID,
			Action:    c.Action,
			Timestamp: c.Timestamp,
			Data:      data,
		})
		ids = append(ids, c.ID)
	}

	if len(records) == 0 {
		return
	}

	if err := e.client.Push(records); err != nil {
		log.Printf("[sync] push error: %v", err)
		return
	}

	e.store.MarkChangesSynced(ids)
	log.Printf("[sync] pushed %d changes", len(records))
}

func (e *Engine) doPull() {
	resp, err := e.client.Pull(e.lastSync.Format(time.RFC3339))
	if err != nil {
		log.Printf("[sync] pull error: %v", err)
		return
	}

	if len(resp.Changes) == 0 {
		return
	}

	merged := 0
	for _, c := range resp.Changes {
		localData, _ := e.store.GetFullRecord(c.TableName, c.RecordID)
		shouldWrite := len(localData) == 0

		if !shouldWrite && len(c.Data) > 0 {
			remoteTime, err1 := time.Parse(time.RFC3339, c.Timestamp)
			var localObj map[string]interface{}
			if err2 := json.Unmarshal(localData, &localObj); err2 == nil {
				if updatedAt, ok := localObj["updated_at"].(string); ok {
					localTime, err3 := time.Parse(time.RFC3339, updatedAt)
					if err1 == nil && err3 == nil && !remoteTime.Before(localTime) {
						shouldWrite = true
					}
				}
			}
		}

		if shouldWrite && len(c.Data) > 0 {
			if err := e.store.PutFullRecord(c.TableName, c.RecordID, c.Data); err != nil {
				log.Printf("[sync] merge error %s/%s: %v", c.TableName, c.RecordID, err)
				continue
			}
			merged++
		}
	}

	if t, err := time.Parse(time.RFC3339, resp.ServerTime); err == nil {
		e.lastSync = t
	}

	log.Printf("[sync] pulled %d changes, merged %d", len(resp.Changes), merged)
}

File: src/lib/sync/engine.ts
import { SyncClient, type SyncRecord } from './client';
import type { DatabaseAdapter, Task, TaskList } from '../adapter/types';

export class SyncEngine {
  private client: SyncClient;
  private lastSync: string;
  private pushTimer: number | null = null;
  private pullTimer: number | null = null;

  constructor(
    private adapter: DatabaseAdapter,
    serverUrl: string,
    token: string,
  ) {
    this.client = new SyncClient(serverUrl, token);
    this.lastSync = new Date().toISOString();
  }

  start(): void {
    this.doPull();
    this.pushTimer = window.setInterval(() => this.doPush(), 3000);
    this.pullTimer = window.setInterval(() => this.doPull(), 30000);
  }

  stop(): void {
    if (this.pushTimer) clearInterval(this.pushTimer);
    if (this.pullTimer) clearInterval(this.pullTimer);
  }

  private async doPush(): Promise<void> {
    try {
      const changes = await (this.adapter as any).getUnsyncedChanges?.();
      if (!changes || changes.length === 0) return;

      const records: SyncRecord[] = changes.map((c: any) => ({
        table_name: c.table_name,
        record_id: c.record_id,
        action: c.action,
        timestamp: c.timestamp,
        data: c.data,
      }));

      await this.client.push(records);
      await (this.adapter as any).markChangesSynced?.(changes);
    } catch (err) {
      console.warn('[sync] push failed:', err);
    }
  }

  private async doPull(): Promise<void> {
    try {
      const resp = await this.client.pull(this.lastSync);
      this.lastSync = resp.server_time;

      let merged = 0;
      for (const c of resp.changes) {
        const existing = await this.getLocalRecord(c.table_name, c.record_id);
        let shouldWrite = !existing;

        if (!shouldWrite && existing && c.data) {
          const remoteTime = new Date(c.timestamp).getTime();
          const localTime = new Date(existing.updated_at).getTime();
          if (!isNaN(remoteTime) && !isNaN(localTime) && remoteTime >= localTime) {
            shouldWrite = true;
          }
        }

        if (shouldWrite && c.data) {
          await this.saveRemoteRecord(c.table_name, c.data);
          merged++;
        }
      }

      if (merged > 0) {
        console.log('[sync] merged', merged, 'of', resp.changes.length, 'changes');
      }
    } catch (err) {
      console.warn('[sync] pull failed:', err);
    }
  }

  private async getLocalRecord(table: string, id: string): Promise<{ updated_at: string } | null> {
    try {
      if (table === 'tasks') {
        const tasks = await this.adapter.getTasks();
        const task = tasks.find((t: Task) => t.id === id);
        return task ? { updated_at: task.updated_at } : null;
      }
      if (table === 'lists') {
        const lists = await this.adapter.getLists();
        const list = lists.find((l: TaskList) => l.id === id);
        return list ? { updated_at: list.updated_at } : null;
      }
    } catch { /* offline or error */ }
    return null;
  }

  private async saveRemoteRecord(table: string, data: any): Promise<void> {
    if (table === 'tasks') {
      await this.adapter.saveTask(data as Task);
    } else if (table === 'lists') {
      await this.adapter.saveList(data as TaskList);
    }
  }
}
